app/javascript/utils/fen-conversion.js

```diff
@@ -0,0 +1,251 @@
+  let halfMoveCounter = 2 * fullMoveCounter + (activeColor === 'w' ? -2 : -1);
+  state[8] = (parseInt(halfMoveCounter / 128));
+  state[9] = (parseInt(halfMoveCounter % 128));
+
+  // set king position
+  state[11] = (blackKing);
+  state[123] = (whiteKing);
+
+  // set color
+  if (activeColor === 'w') {
+    state[56] = (1);
+  } else {
+    state[56] = (-1);
+  }
+
+  // init for castling
+  state[78] = (-1);
+  state[79] = (-1);
+  state[62] = (-1);
+  state[63] = (-1);
+
+  // change state if castling is possible
+  for (var k = 0; k < castling.length; k++) {
+    // white right - kleine rochade für weiß
+    if (castling[k] === 'K') {
+      state[79] = 0;
+    }
+    // white left - große rochade für weiß
+    else if (castling[k] === 'Q') {
+      state[78] = 0;
+    }
+    // black right - kleine rochade für schwarz
+    else if (castling[k] === 'k') {
+      state[63] = 0;
+    }
+    // black left - große rochade für schwarz
+    else if (castling[k] === 'q') {
+      state[62] = 0;
+    }
+  }
+
+  // set enpassant
+  let mapping = generateMapping();
+  state[61] = mapping.toBackend[enPassant];
+  state[77] = mapping.toBackend[enPassant];
+
+  return state;
+}
+
+export function generateFen(state) {
+  let skip = 0, fen = '', zero = 0, toPiece = generatePieceMapping();
+
+  for (var i = 0; i < state.length; i++) {
+    // field is empty
+    if (state[i] === 0) {
+      zero += 1;
+    }
+    // field contains a piece
+    else {
+
+      // before concatinate piece to fen, check if zeros exist
+      if (zero > 0) {
+        fen += zero;
+        zero = 0;
+      }
+      fen += toPiece[state[i]];
+    }
+
+    skip++;
+
+    // shadow board
+    if (skip === 8) {
+
+      // concatinate rest to fen if exists
+      if (zero > 0) {
+        fen += zero;
+        zero = 0;
+      }
+
+      // concatinate '/'
+      if (i < 118) {
+        fen += '/';
+      }
+
+      // skip shadow board and reset skip
+      i += 8;
+      skip = 0;
+    }
+  }
+
+  // set current player
+  if (state[56] === 1) {
+    // white
+    fen += ' w ';
+  } else {
+    // black
+    fen += ' b ';
+  }
+
+  // set Rochade
+  if (state[79] === 0 || state[78] === 0 ||
+    state[62] === 0 || state[63] === 0) {
+    if (state[79] === 0) {
+      fen += 'K';
+    }
+    if (state[78] === 0) {
+      fen += 'Q';
+    }
+    if (state[62] === 0) {
+      fen += 'k';
+    }
+    if (state[63] === 0) {
+      fen += 'q';
+    }
+  } else {
+    fen += '-';
+  }
+
+  // set En passant
+  if (state[61] > 0 || state[77] > 0) {
+    let position = generateMapping();
+    if (state[61] > 0) {
+      fen += ' ' + position.toFrontend[state[61]];
+    }
+    else if (state[77] > 0) {
+      fen += ' ' + position.toFrontend[state[77]];
+    }
+  } else {
+    fen += ' -';
+  }
+
+  // set clock for halfmoves since last capture or pawn advance
+  // not implemented, so always 0
+  fen += ' 0 ';
+
+  // set fullmove number
+  let halfMoveCounter = state[8] << 7 + state[9];
```

Comment:
@graup: this new line instead of: `let halfMoveCounter = 128 * state[8] + state[9];` causes problems. Finally found this bug. Normally, we should not try to 'optimize' working functions while refactoring, since new introduced bugs are very hard to trace ;)